### PR TITLE
CLI-657: Improve SSH key polling and permission checks

### LINKED ISSUE #818: CLI-657: Check whether the user role has permissions to SSH into a non production environment and fail gracefully if not
Cloud IDE users can run into the following issue:

Screen Shot 2022-02-17 at 08 53 25

This is due to roles (e.g. [Developer role by default](https://docs.acquia.com/cloud-platform/access/teams/permissions/default/)) not having permissions to create SSH keys.

Instead of just timing out, we should first perform a permissions check and fail gracefully with instructions.

### PR DESCRIPTION
**Pull request by danepowell**
**Motivation**

Fixes #818 

**Proposed changes**

- Add permission checks for git, prod, and non-prod environments (the three places where SSH keys can be installed)
- Warn the user if they lack permissions for any of these
- Poll all three places to make sure SSH keys are fully installed and available

**Testing steps**

0. Use or mimic an IDE (export REMOTEIDE_UUID and AH_SITE_ENVIRONMENT=ide)
1. Follow the [contribution guide](https://github.com/acquia/cli/blob/master/CONTRIBUTING.md#building-and-testing) to set up your development environment.
2. Clear the kernel cache to pick up new and changed commands: `./bin/acli ckc`
3. Remove any existing local IDE SSH keys
4. Run `acli ide:wizard`
5. Depending on whether you have permissions to access git / prod / non-prod environments in the respective application, see that the corresponding environment is either polled or you are warned that you lack access.

**Merge requirements**
- [x] _Bug_, _enhancement_, or _breaking change_ label applied
- [ ] Manual testing by a reviewer

### CODE REVIEW

File: src/Command/Ide/Wizard/IdeWizardCommandBase.php [hunk 1]
```diff
@@ -71,9 +71,9 @@ protected function getSshKeyLabel() {
   /**
    * @throws \Acquia\Cli\Exception\AcquiaCliException
    */
-  protected function deleteThisSshKeyFromCloud(): void {
+  protected function deleteThisSshKeyFromCloud($output): void {
```
Comment (danepowell) [inline src/Command/Ide/Wizard/IdeWizardCommandBase.php:74]: This fixes a latent bug (missing parameter) that prevents you from deleting SSH keys on cloud. We need to make sure it gets committed regardless of what happens to this PR.

File: src/Command/Ssh/SshKeyCommandBase.php [hunk 2]
```diff
@@ -151,33 +148,99 @@ protected function pollAcquiaCloudUntilSshSuccess(
...
+    $vcs_url = $environment_nonprod = $environment_prod = NULL;
+    if (in_array('add ssh key to git', $perms, TRUE)) {
+      $poll_git = TRUE;
+      $full_url = $this->getAnyVcsUrl($cloud_app_uuid);
+      $url_parts = explode(':', $full_url);
+      $vcs_url = $url_parts[0];
+    }
+    else {
+      $poll_git = FALSE;
+      $output->writeln('<comment>You do not have access to Cloud Platform git on this application and will not be able to clone your codebase to this IDE. Check that you have the <options=bold>add ssh key to git</> permission. Documentation on Cloud Teams permissions: <href=https://docs.acquia.com/cloud-platform/access/teams/permissions/default/>https://docs.acquia.com/cloud-platform/access/teams/permissions/default/</>');
```
Comment (grasmash) [inline src/Command/Ssh/SshKeyCommandBase.php:166]: Let's use `$this->io`

File: src/Command/Ssh/SshKeyCommandBase.php [hunk 3]
```diff
@@ -151,33 +148,99 @@ protected function pollAcquiaCloudUntilSshSuccess(
...
+      $poll_nonprod = FALSE;
+      $output->writeln('<comment>You do not have access to Cloud Platform non-prod environments on this application and will not be able to clone your non-prod sites to this IDE. Check that you have the <options=bold>add ssh key to non-prod environments</> permission. Documentation on Cloud Teams permissions: <href=https://docs.acquia.com/cloud-platform/access/teams/permissions/default/>https://docs.acquia.com/cloud-platform/access/teams/permissions/default/</>');
+    }
+    if (in_array('add ssh key to prod', $perms, TRUE)) {
+      $poll_prod = TRUE;
+      $environment_prod = $this->getAnyProdAhEnvironment($cloud_app_uuid);
+    }
+    else {
+      $poll_prod = FALSE;
+      $output->writeln('<comment>You do not have access to Cloud Platform prod environments on this application and will not be able to clone your prod sites to this IDE. Check that you have the <options=bold>add ssh key to prod environments</> permission. Documentation on Cloud Teams permissions: <href=https://docs.acquia.com/cloud-platform/access/teams/permissions/default/>https://docs.acquia.com/cloud-platform/access/teams/permissions/default/</>');
```
Comment (grasmash) [inline src/Command/Ssh/SshKeyCommandBase.php:182]:
```suggestion
      $output->writeln('<comment>You do not have access to Cloud Platform <options=bold>prod</> environments on this application and will not be able to clone your prod sites to this IDE. Check that you have the <options=bold>add ssh key to prod environments</> permission. Documentation on Cloud Teams permissions: <href=https://docs.acquia.com/cloud-platform/access/teams/permissions/default/>https://docs.acquia.com/cloud-platform/access/teams/permissions/default/</>');
```

File: src/Command/Ssh/SshKeyCommandBase.php [hunk 4]
```diff
@@ -151,33 +148,99 @@ protected function pollAcquiaCloudUntilSshSuccess(
...
     $loop = Loop::get();
-    $spinner = LoopHelper::addSpinnerToLoop($loop, 'Waiting for the key to become available on the Cloud Platform', $output);
-
-    // Wait for SSH key to be available on a web.
+    $spinner_git = LoopHelper::addSpinnerToLoop($loop, 'Waiting for the key to become available in Cloud Platform git', $output);
+    $spinner_nonprod = LoopHelper::addSpinnerToLoop($loop, 'Waiting for the key to become available in Cloud Platform non-prod environments', $output);
+    $spinner_prod = LoopHelper::addSpinnerToLoop($loop, 'Waiting for the key to become available in Cloud Platform prod environments', $output);
     $cloud_app_uuid = $this->determineCloudApplication(TRUE);
-    $environment = $this->getAnyNonProdAhEnvironment($cloud_app_uuid);
+    $permissions = $this->cloudApiClientService->getClient()->request('get', "/applications/{$cloud_app_uuid}/permissions");
```
Comment (grasmash) [inline src/Command/Ssh/SshKeyCommandBase.php:155]: Seems like these permissions checks belong in a separate method.

File: src/Command/Ssh/SshKeyCommandBase.php (hunk 2 above)
Comment (danepowell) [reply src/Command/Ssh/SshKeyCommandBase.php:166]: `$this->io` seems really broken. Depending on the method (warning vs comment, for instance), it won't respect inline styles (`<options=bold>`). And it keeps breaking the URL into multiple lines for no apparent reason.

File: src/Command/Ssh/SshKeyCommandBase.php [hunk 5]
```diff
@@ -151,33 +148,99 @@ protected function pollAcquiaCloudUntilSshSuccess(
...
 
     // Poll Cloud every 5 seconds.
-    $loop->addPeriodicTimer(5, function () use ($output, $loop, $environment, $spinner) {
-      try {
-        $process = $this->sshHelper->executeCommand($environment, ['ls'], FALSE);
-        if ($process->isSuccessful()) {
-          LoopHelper::finishSpinner($spinner);
-          $loop->stop();
-          $output->writeln("\n<info>Your SSH key is ready for use!</info>\n");
+    $loop->addPeriodicTimer(5, function () use ($output, $loop, $spinner_git, $spinner_nonprod, $spinner_prod, &$poll_git, &$poll_nonprod, &$poll_prod, $vcs_url, $environment_nonprod, $environment_prod) {
```
[the remark this answers is not in the document]
Comment (danepowell) [reply src/Command/Ssh/SshKeyCommandBase.php:186]: Tell me about it... the problem is each environment is just unique enough to make this really hard to do DRY without putting closures inside closures inside closures, or adding a bunch of conditionals.

I'll see what I can do.

File: src/Command/Ssh/SshKeyCommandBase.php (hunk 2 above)
Comment (grasmash) [reply src/Command/Ssh/SshKeyCommandBase.php]: Yeah. That's the downside of it. It basically takes over all formatting. That's why I've ended up mixing it with `$output->writeln()`.

**Pull request merged**